apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js

```diff
@@ -33,16 +33,22 @@ const {
 class PageTemplateModal extends Component {
 	state = {
 		isLoading: false,
-		slug: '',
-		title: '',
-		blocks: {},
+		previewedTemplate: null,
+		blocksBySlug: {},
+		templatesBySlug: {},
 		error: null,
 		isOpen: false,
 	};
 
 	constructor( props ) {
 		super();
-		this.state.isOpen = ! isEmpty( props.templates );
+		const hasTemplates = ! isEmpty( props.templates );
+		this.state.isOpen = hasTemplates;
+		// Select the first template automatically.
+		if ( hasTemplates ) {
+			this.state.previewedTemplate = get( props.templates, [ 0, 'slug' ] );
+			this.state.templatesBySlug = keyBy( props.templates, 'slug' );
```

Comment:
Do you think that it's completely needed to populate the state with the `templatesBySlug` object? It seems to be used only to get the template title. What about...

```es6
	getTitleBySlug( slug = this.state.previewedTemplate ) {
		return get( find( this.props.templates, { slug } ), 'title', '' );
	}
```

Reply:
While the `find` solution definitely works, it requires to always loop over the array for each call of that function which we do twice per render. Having the dictionary in state created by `keyBy` makes this loop happen once per page view and any further lookup is instantaneous while keeping a really low memory impact (remember objects are not copies but references here). 

The only disadvantage is that `templatesBySlug` won't update if we get re-rendered with a different set of `templates` but this is true for our whole plugin so I didn't bother at this point (also because that situation is not possible to be triggered with our current code).

Reply:
Ok, got it. Already getting into very detailed issues, make sense just store the title instead of the whole template object? My concern provably unfounded is filling the state with duplicated data.
Take this comment much as a teaching session rather than a feedback request. ;-)

Reply:
We are definitely in a theoretical area here but I like it! Normally, I would have implemented all this props/state mess as a Redux store and in those, it is a well established pattern to use two data structs to store data like this:

- `templatesBySlug` - an object keyed by the slug with values of the full template objects. an example:

```
templatesBySlug = {
  blank: { title: 'Blank', … },
  about: { title: 'About', … },
  …
}
```

- and the other part `templateList` - an array of slugs that are supposed to be displayed (also explicitly stating their order). an example:

```
templateList = [ 'blank', 'about', 'contact', … ]
```

I think we are slowly getting there and I expect us to switch to Redux at some point (with @wordpress/data) so just keeping it here for an inspiration.

Regarding storing just titles, the difference is that those would be stored as copies in memory (doubling the usage) as that's what JS does for scalar types. On the other hand, when storing objects, they are not copies but only references.

In case of titles, the memory impact is extremely tiny even for copies (they are short strings and there is a small number of them). I'm happy to switch to that approach for now as we don't access anything other than titles.

Reply:
I really appreciate it. :thanks: